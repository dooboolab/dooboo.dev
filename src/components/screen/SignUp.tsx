import React, { Component } from 'react';
import styled from 'styled-components';

const Container = styled.div`
  flex: 1;
  background-color: transparent;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

interface IProps {
  history?: any;
}

function Page(props: IProps) {
  return (
    <Container>
      <div
        data-testid='myText'
      >Sign Up</div>
    </Container>
  );
}

export default Page;
