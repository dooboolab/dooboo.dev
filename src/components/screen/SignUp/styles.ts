import { Form } from 'formik';
import styled from 'styled-components';

type TInputProps = {
  'data-testid'?: string;
  widthinrem?: number;
  right?: boolean;
}

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  background: ${(props) => props.theme.background};  
  height: 100%;
  justify-content: center;
  align-items: center;
`;

export const FormInnerWrapper = styled.div`
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin-bottom: 2rem;
`;

export const SignContainer = styled(Form)<TInputProps>`
  display: flex;
  width: ${({ widthinrem }) => widthinrem}rem;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
`;

export const BottomTextBtnContainer = styled.div`
  display: flex;
  flex-direction: row;
  margin-top: 1rem;
  justify-content: center;
  align-items: center;
`;

export const BottomTextButton = styled.input.attrs(() => ({
  type: 'button',
}))<TInputProps>`
  margin-left: ${({ right }) => (right ? '1rem' : 0)};
  border: none;
  text-decoration-line: underline;
  background-color: transparent;
  color: ${({ theme }) => theme.highlighted};
  cursor: pointer;
`;
