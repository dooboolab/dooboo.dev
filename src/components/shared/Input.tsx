import React, { useEffect, memo, NamedExoticComponent } from 'react';
import { ErrorMessage } from 'formik';
import styled from 'styled-components';
import { FormLabel, FormError, FormInput } from '../ui/Forms';

export const FormContainer = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 4rem;
  align-self: flex-end;
`;

export const FormInnerContainer = styled.div`
  display: flex;
  flex-direction: row-reverse;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
  margin: 1rem auto 0.4rem auto;
`

export const getLabelCapitalize = (label: string) => `${label.charAt(0).toUpperCase()}${label.slice(1)}`;

export const SimpleFormInput: NamedExoticComponent<any> = memo(({ field, form, type, width, receivedEmail, setEmail }) => {
  const { name, ...restFields } = field;
  const  { touched: touches, errors, setFieldValue } = form;
  const isTouched = touches && touches[name];
  const error = errors && errors[name]
  const rest = { type, width, isTouched, error, ...restFields }
  useEffect(() => {
    if (name === 'email' && setEmail && receivedEmail) {
      setFieldValue(name, receivedEmail);
    }
    return () => {
      if (setEmail) {
        setEmail('');
      }      
    }
  }, [setEmail, receivedEmail]);
  
  return (
    <FormContainer>
      <FormInnerContainer>        
        <FormInput id={name} {...rest} />
        <FormLabel data-testid={`label-${name}`} htmlFor={name}>{(name && name.length) && getLabelCapitalize(name)}</FormLabel>
      </FormInnerContainer>
      <ErrorMessage name={name}>
        {err => <FormError data-testid={`errors-${name}`}>{err}</FormError>}
      </ErrorMessage>
    </FormContainer>
  )
})