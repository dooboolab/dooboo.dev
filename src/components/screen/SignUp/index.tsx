import React, { useContext, useCallback } from 'react';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import Swal from 'sweetalert2';
import withReactContent from 'sweetalert2-react-content';

import { AppContext } from '../../../contexts';
import {
  Wrapper,
  SignContainer,
  FormInnerWrapper,
  BottomTextBtnContainer,
  BottomTextButton,
} from './styles';
import { SimpleFormInput } from '../../shared/Input';
import Button from '../../shared/Button';

const ReactSwal = withReactContent(Swal);

export const signUpList = [{
  label: 'email',
  type: 'text',
}, {
  label: 'password1',
  type: 'password',
}, {
  label: 'password2',
  type: 'password',
}];

const signUpValidationSchema = Yup.object({
  email: Yup.string().email(),
  password1: Yup.string().min(6, 'password must be at least 6 characters.').required('password is required'),
  password2: Yup.string().min(6, 'password must be at least 6 characters.').oneOf([Yup.ref('password1'), null], 'passwords must match'),
});

const onSignUpSubmit = ({ email, password1 }, { setSubmitting, resetForm }, history, createUserWithEmailAndPW) => {
  if (createUserWithEmailAndPW) {
    createUserWithEmailAndPW(email, password1)
      .then(({ user: { email } }) => {
        setSubmitting(false);
        ReactSwal.fire({
          title: 'Signing up completed.',
          html: `${email} has been registered in our system.<br />Please sign in now.`,
          type: 'success',
          heightAuto: false,
          customClass: { icon: 'swal-custom-overflow' },
        }).then(() => {
          history.push({
            pathname: '/signin',
            state: { email },
          });
        });
      })
      .catch(({ code, message }) => {
        ReactSwal.fire({
          title: code,
          text: message,
          type: 'error',
          heightAuto: false,
        });
        resetForm();
      });
  }
};

export default ({ history }) => {
  const { state: { firebase: { createUserWithEmailAndPW } } } = useContext(AppContext);
  const onClickSignIn = useCallback(() => history.push('/signin'), []);

  return (
    <Wrapper>
      <Formik
        initialValues={{
          email: '',
          password1: '',
          password2: '',
        }}
        validationSchema={signUpValidationSchema}
        onSubmit={(values, actions) => onSignUpSubmit(values, actions, history, createUserWithEmailAndPW)}
      >
        {({
          isSubmitting,
          isValid,
        }) => (
          <SignContainer widthinrem={20}>
            <FormInnerWrapper>
              {signUpList.map(({ label, type }) => (
                <Field
                  key={label}
                  name={label}
                  type={type}
                  width={7}
                  component={SimpleFormInput}
                />
              ))}
            </FormInnerWrapper>
            <Button
              type="submit"
              disabled={isSubmitting || !isValid}
              isLoading={isSubmitting}
              text="Sign Up"
              height="3.5rem"
            />
            <BottomTextBtnContainer>
              <BottomTextButton
                onClick={onClickSignIn}
                value="Sign In"
              />
            </BottomTextBtnContainer>
          </SignContainer>
        )}
      </Formik>
    </Wrapper>
  );
};
