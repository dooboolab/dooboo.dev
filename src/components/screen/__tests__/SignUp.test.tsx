import React from 'react';
import { withRouter } from 'react-router';

import SignUp, { signUpList } from '../SignUp';
import { getLabelCapitalize } from '../../shared/Input';
import { history, within, render, fireEvent, wait, waitForElement } from '../../../../test/test-utils';

const props = {
  history,
};

describe('[SignUp] render', () => {
  it('renders without crashing', () => {
    const component: any = render(<SignUp {...props} />);
    expect(component.firstChild).toMatchSnapshot();
  });
});

describe('[SignUp] Interaction', () => {
  let renderResult: any;

  const LocationDisplay = withRouter(({ location: { pathname, state } }) => {
    let email = '';
    if (state) {
      ({ email } = state);
    }
    return (
      <>
        <div data-testid="location-display">{pathname}</div>
        <span data-testid="location-state-email">{email}</span>
      </>
    );
  });

  beforeEach(() => {
    renderResult = render(
      <div>
        <SignUp {...props} />
        <LocationDisplay />
      </div>,
      { route: '/signup' }
    );
  });

  it('should simulate [onClick] when [btn] has been clicked', () => {
    const { getByText } = renderResult;
    const spy = jest.fn();
    const SignUpSubmitBtn = getByText('Sign Up').closest('button');
    SignUpSubmitBtn.addEventListener('click', spy,);
    fireEvent.click(SignUpSubmitBtn);
    expect(spy).toHaveBeenCalled();
  });

  it('testing moving to the home when successfully signed up.', async () => {
    window.scrollTo = () => {}; // for requirement of Swal2 test
    const { container, debug, getByLabelText, getByText, getByTestId } = renderResult;
    const wrapper = container.parentNode;
    const emailInput = getByLabelText(getLabelCapitalize(signUpList[0].label));
    const pwInput = getByLabelText(getLabelCapitalize(signUpList[1].label));
    const pwInput2 = getByLabelText(getLabelCapitalize(signUpList[2].label));
    const submitBtn = getByText('Sign Up').closest('button');

    fireEvent.change(emailInput, { target: { value: 'hello@signup.com' } });
    fireEvent.blur(emailInput);
    fireEvent.change(pwInput, { target: { value: 'hello@signup.com' } });
    fireEvent.blur(pwInput);
    fireEvent.change(pwInput2, { target: { value: 'hello@signup.com' } });
    fireEvent.blur(pwInput2);
    fireEvent.click(submitBtn);

    const Swal2 = await waitForElement(() => wrapper.querySelector('.swal2-container'));
    const Swal2OKBtn = within(Swal2).getByText('OK');

    fireEvent.click(Swal2OKBtn);

    const locationComp = await waitForElement(() => getByTestId('location-display'));
    const locationStateEmail = await waitForElement(() => getByTestId('location-state-email'));
    expect(locationComp.textContent).toBe('/signin');
    expect(locationStateEmail.textContent).toBe('hello@signup.com');
  });

  /* it('see what happens when click "Sign In" button', async () => {
    window.scrollTo = () => {}; // for requirement of Swal2 test
    const { container, getByText } = renderResult;
    const forgotPWBtn = getByText('Forgot password?');
    const wrapper = await waitForElement(() => container.parentNode);
    fireEvent.click(forgotPWBtn);
    const Swal2 = wrapper.querySelector('.swal2-container');
    const swal2EmailInput = Swal2.querySelector('input[type="email"]');
    const swal2OKBtn = Swal2.querySelector('button.swal2-confirm');
    fireEvent.change(swal2EmailInput, { target: { value: 'test@test.com' }});
    fireEvent.click(swal2OKBtn);
    await wait(() => expect(wrapper.querySelector('#swal2-content').textContent).toMatch('a link for reset your password has been sent.'));

    const { container, debug, getByText } = renderResult;
    const moveToSignInBtn = getByText('Sign In');
    debug(moveToSignInBtn);
  }); */
});
