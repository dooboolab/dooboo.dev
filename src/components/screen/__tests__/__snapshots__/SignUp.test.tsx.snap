// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SignUp] render renders without crashing 1`] = `undefined`;
